### PR TITLE
 Add a doc for cmake

### PR DESCRIPTION
**Pull request by snnn**
**Description**: 
Add a document for our cmake files

**Motivation and Context**
- Why is this change required? What problem does it solve?
In cmake world, it's usually one thing could be done in two different ways. Either should work. But mixing them is not good.  So, let's define, which one we want to use. 

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by pranavsharma** [comment]
Thanks. Can you please add a reason for each guideline? or at least link to a document that explains the reason?

**Comment by snnn** [comment]
> Thanks. Can you please add a reason for each guideline? or at least link to a document that explains the reason?

I guess I have explained most of them. Please let me know which are not clear. I'll add materials to them.

File: docs/cmake_guideline.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ target_link_libraries(your_exe PRIVATE lib1)
 ```
 If you remove it, the code may still work. But, please don't omit it.
 
-Use PRIVATE if you don't know which to use. If you don't understand why it's is needed, or what's the differece between the three, that's fine. Just always put PRIVATE there.
+Use PRIVATE if you don't know which to use. If you don't understand why it's is needed, or what's the difference between the three, that's fine. Just always put PRIVATE there.
```
Comment (pranavsharma) [inline docs/cmake_guideline.md]: what's the recommendation for public and interface usage?

File: docs/cmake_guideline.md [hunk 2]
```diff
@@ -36,29 +36,29 @@ Not
...
 On Linux, it matters when one static library references another.
 
-So, in general, please always put them in right order(according to their dependency relationship).
+So, in general, please always put them in right order (according to their dependency relationship).
 
 # No circular dependency between libs
-If A depends on B, then B shouldn't depends on A.  This rule is strict. Don't ask for an exception.
+If A depends on B, then B shouldn't depend on A.  Don't ask for an exception. Let's just keep it simple.
 
 # Don't call target\_link\_libraries on static libraries
```
Comment (pranavsharma) [inline docs/cmake_guideline.md]: why?

File: docs/cmake_guideline.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+target_link_libraries(your_exe PRIVATE lib1)
+```
+If you remove it, the code may still work. But, please don't omit it.
+
+Use PRIVATE if you don't know which to use. If you don't understand why it's is needed, or what's the difference between the three, that's fine. Just always put PRIVATE there.
+
+This rule also applies to "target\_include\_directories".
+
+# Static library order matters
+First, you should know, when you link static libraries to an executable(or share library) target, the order matters a lot.
```
Comment (pranavsharma) [inline docs/cmake_guideline.md]: typo: shared

File: docs/cmake_guideline.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+However, even you told compiler to link with pthread, sometimes it doesn't listen to you. It may ignore your order, mess your life. see [https://github.com/protocolbuffers/protobuf/issues/5923](https://github.com/protocolbuffers/protobuf/issues/5923). So, do the double check. When the build is done, please manually check the result binary if pthread is there. (Run "ldd" command and grep pthread).
+
+Another related thing is: if std::atomic was in use, please also add the atomic lib there. Because some uses of std::atomic require linking to libatomic. see [https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_concurrency.html](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_concurrency.html)
+
+# Don't use the "-pthread" flags directly. 
+Because:
+1. It doesn't work with nvcc(the CUDA compiler)
+2. Not portable. 
+
+Don't bother to add this flag to your compile time flags. On Linux, it's useless. On some very old unix-like system, it may be helpful, but we don't care.
```
Comment (pranavsharma) [inline docs/cmake_guideline.md]: we don't care -> we only support Ubuntu 16.04.

File: docs/cmake_guideline.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+- First you should know ,there is no link step for a static lib
+- Second, once you hit the ordering problem, it would be harder to fix. Because many of the deps were implicit, and their position would be out of our control.
+
+# Every linux program(and shared lib) should link to libpthread and libatomic
+In Linux world, there are two set of pthread symbols. A fake one in the standard c library, and a real one in pthread.so. If the real one is not loaded while the process was starting up, then the process shouldn't use multiple threading because the core part was missing.
+
+So, We append "Threads::Threads" to the lib list of every shared lib(*.so,*.dll) and exe target. It's easy to get missed. If it happened, the behavior is undefined.
+
+However, even you told compiler to link with pthread, sometimes it doesn't listen to you. It may ignore your order, mess your life. see [https://github.com/protocolbuffers/protobuf/issues/5923](https://github.com/protocolbuffers/protobuf/issues/5923). So, do the double check. When the build is done, please manually check the result binary if pthread is there. (Run "ldd" command and grep pthread).
```
Comment (pranavsharma) [inline docs/cmake_guideline.md]: mess your life -> cause issues

**Comment by jywu-msft** [comment]
kicked off failing CI due to out of disk space on c drive.

**Pull request merged**